Adding display of queued speakers in group chat windows.

Esalarc opened the pull request:
- Adding display of queued speakers in group chat windows.
 - all drafted characters are highlighted - current in --active, others in --golden
 - Position in queue displayed next to each drafted character's name 

![image](https://github.com/SillyTavern/SillyTavern/assets/12687033/f707b176-fd4f-4784-b3fc-47fb33aa51c7)


<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

Esalarc committed:
- Adding display of queued speakers in group chat windows.

Wolfsblvt commented on public/scripts/group-chats.js:
```diff
@@ -1309,11 +1330,13 @@ function getGroupCharacterBlock(character) {
+    template.toggleClass('is_active', character.queueOrder === 1);
+    template.toggleClass('is_queued', character.queueOrder > 1);
     template.toggleClass('disabled', isGroupMemberDisabled(character.avatar));
+    template.toggleClass('is_active', character.avatar === (this_chid && characters[this_chid] && characters[this_chid].avatar));
```
Why are you toggling `is_active` class twice here? Which one is the correct? Should likely be merged.

Esalarc replied:
Because I'm not very smart :)  Fixed in latest commit

Wolfsblvt commented on public/scripts/group-chats.js:
```diff
@@ -1309,11 +1330,13 @@ function getGroupCharacterBlock(character) {
     template.find('.avatar img').attr({ 'src': avatar, 'title': character.avatar });
-    template.find('.ch_name').text(character.name);
-    template.attr('chid', characters.indexOf(character));
+    template.find('.ch_name').text(character.name + (character.queueOrder > 0?'   (#' + character.queueOrder + ')':''));    template.attr('chid', characters.indexOf(character));
```
Would prefer it if there would be another div/span block for the queue number. Makes removing, hiding and styling them easier. And some extensions/or functions sometimes query `.ch_name` to get the display name.

I think this should be easily achievable by modfying the template itself in HTML.

Esalarc replied:
Done

Wolfsblvt commented on public/scripts/group-chats.js:
```diff
@@ -857,14 +859,21 @@ async function generateGroupWrapper(by_auto_mode, type = null, params = {}) {
-
+        if (showChatQueue){
+            for (let i = 0; i < activatedMembers.length; ++i){
+                characters[activatedMembers[i]].queueOrder = (i+1);
```
I would suggest storing the queue order in some other place than the character object.
The character is a global representation of the char card, and is not related to any chat really.  
This might give the wrong impression that this data is directly related to the character, and not the group member in the context of this group chat.

You can either think about just using a local Map inside the `group-chats.js` file, or maybe even saving the queue order into the chat context, if you want it persisted through possible reloads of the page. Though I think the latter doesn't make sense, because with a refresh generation stops anyway, right?

Cohee1207 replied:
This should not be leaving the generateGroupWrapper function even. By keeping a local map of the draft and passing it to printGroupMembers (default value = empty map) it is possible to keep the JSON data clean and not create any more global variables.

Esalarc replied:
Removed it from the character object.  I wasn't able to keep it local to generateGroupWrapper though.  There's an issue there when you pop out the chat members while a response is being generated.  The pop out runs through printGroupMembers, but knew nothing about the queue order determined during the draft so highlighting did not function.

Cohee1207 commented:
I'd like it to be an option, please. That is not a bad addition, but I figured for some it may be a distraction or immersion breaker.

Wolfsblvt commented on public/scripts/group-chats.js:
```diff
@@ -785,7 +786,8 @@ async function generateGroupWrapper(by_auto_mode, type = null, params = {}) {
     }
-
+    
+    const showChatQueue = (!(typingIndicator.length === 0 && !isStreamingEnabled()) && openGroupId);
```
Is there a reason you explicitly excluded streaming from displaying the queue order?
I am not using group chats a lot, so I might be missing something here.
But wouldn't generally having that indicator, even if the first char in the queue is streaming its response right now, still help?

Cohee1207 replied:
Agreed that streaming should not be a problem here since the group draft loop actually waits for the stream to complete before proceeding to the next member. So calling a refresh list at the beginning of every iteration should be possible

Esalarc replied:
I was trying to avoid conflicting with the typing indicator (code around line 799)...with the addition of an on/off toggle in latest commit, that concern goes away, so I've removed that code.

Wolfsblvt reviewed (changes requested):
Have some minor changes/questions about this. Nothing a dealbreaker.

Besides what Cohee said, a toggle option for this would be kinda important.

Generally, I really like the addition. Cool idea!

Cohee1207 commented on public/scripts/group-chats.js:
```diff
@@ -1309,11 +1330,13 @@ function getGroupCharacterBlock(character) {
     template.find('.avatar img').attr({ 'src': avatar, 'title': character.avatar });
-    template.find('.ch_name').text(character.name);
-    template.attr('chid', characters.indexOf(character));
+    template.find('.ch_name').text(character.name + (character.queueOrder > 0?'   (#' + character.queueOrder + ')':''));    template.attr('chid', characters.indexOf(character));
```
The pound sign could be a part of the CSS style added by with ::before selector instead of a JS-coded value. That would ofc. require it to be a separate element in the name row.

Esalarc replied:
css is the spawn of satan, so I had avoided that.  Moved in latest commit.

Esalarc committed:
Group Chat Queue Display
- removing queue position from character object - storing in map in group_chat.js instead
- moving queue position to div
- moving '#' from js to css ::before
- adding option to settings
- cleaning up guards

Esalarc commented:
> I'd like it to be an option, please. That is not a bad addition, but I figured for some it may be a distraction or immersion breaker.

Added an option in latest commit

Cohee1207 committed:
Fix i18n attributes

Cohee1207 committed:
Fix char data typedef

Cohee1207 committed:
Fix code formatting

Cohee1207 committed:
Replace border with outline

Cohee1207 committed:
Reset queue order when opening another group chat

Cohee1207 committed:
Remove old field reference

Cohee1207 committed:
Reduce font-size of position value

Cohee1207 reviewed (approved):
Tested and adjusted some visuals. Thank you.
